File: docker-mastery-multitrack/02-language-quickstart/python/src/main.py
"""Single-process, in-memory teaching API. Restarting loses all tasks."""
from datetime import datetime, timezone
from uuid import uuid4

from fastapi import FastAPI, HTTPException, Response
from pydantic import BaseModel, ConfigDict, Field, field_validator

app = FastAPI(title="Task API", version="1.0.0")


class TaskCreate(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid")
    title: str = Field(min_length=1, max_length=255)
    description: str = Field(default="", max_length=2000)

    @field_validator("title")
    @classmethod
    def nonblank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("title must not be blank")
        return value


class TaskReplace(TaskCreate):
    completed: bool


class Task(TaskReplace):
    id: str
    created_at: datetime
    updated_at: datetime


tasks: dict[str, Task] = {}


@app.get("/health")
async def health():
    return {"status": "healthy", "version": "1.0.0", "storage": "memory"}


@app.get("/api/tasks")
async def list_tasks():
    return {"tasks": list(tasks.values()), "total": len(tasks)}


@app.post("/api/tasks", status_code=201)
async def create_task(body: TaskCreate):
    now = datetime.now(timezone.utc)
    task = Task(**body.model_dump(), completed=False, id=str(uuid4()),
                created_at=now, updated_at=now)
    tasks[task.id] = task
    return task


def lookup(task_id: str) -> Task:
    if task_id not in tasks:
        raise HTTPException(404, "Task not found")
    return tasks[task_id]


@app.get("/api/tasks/{task_id}")
async def get_task(task_id: str):
    return lookup(task_id)


@app.put("/api/tasks/{task_id}")
async def replace_task(task_id: str, body: TaskReplace):
    previous = lookup(task_id)
    task = Task(**body.model_dump(), id=task_id, created_at=previous.created_at,
                updated_at=datetime.now(timezone.utc))
    tasks[task_id] = task
    return task


@app.delete("/api/tasks/{task_id}", status_code=204)
async def delete_task(task_id: str):
    lookup(task_id)
    del tasks[task_id]
    return Response(status_code=204)


@app.get("/metrics")
async def metrics():
    completed = sum(task.completed for task in tasks.values())
    values = {"task_count": len(tasks), "task_completed_count": completed,
              "task_pending_count": len(tasks) - completed}
    body = "".join(f"# TYPE {name} gauge\n{name} {value}\n"
                   for name, value in values.items())
    return Response(body, media_type="text/plain; version=0.0.4")


@app.get("/")
async def root():
    return {"message": "Task API", "version": "1.0.0", "docker_track": "python",
            "endpoints": {"health": "/health", "tasks": "/api/tasks", "metrics": "/metrics"}}

File: docker-mastery-multitrack/02-language-quickstart/rust/src/main.rs
//! Single-process, in-memory teaching API. Restarting loses all tasks.
use actix_web::{error, middleware::Logger, web, App, HttpResponse, HttpServer, Result};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, sync::{Mutex, MutexGuard}};
use uuid::Uuid;

#[derive(Clone, Serialize)]
struct Task {
    id: String,
    title: String,
    description: String,
    completed: bool,
    created_at: DateTime<Utc>,
    updated_at: DateTime<Utc>,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct TaskCreate {
    title: String,
    #[serde(default)]
    description: String,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct TaskReplace {
    title: String,
    #[serde(default)]
    description: String,
    completed: bool,
}

type Store = Mutex<HashMap<String, Task>>;
fn lock(data: &web::Data<Store>) -> Result<MutexGuard<'_, HashMap<String, Task>>> {
    data.lock().map_err(|_| error::ErrorInternalServerError("Storage unavailable"))
}
fn validate(title: &str, description: &str) -> Result<()> {
    if title.trim().is_empty() || title.chars().count() > 255 || description.chars().count() > 2000 {
        return Err(error::ErrorBadRequest("Invalid title or description length"));
    }
    Ok(())
}
async fn health() -> HttpResponse {
    HttpResponse::Ok().json(serde_json::json!({"status":"healthy", "version":"1.0.0", "storage":"memory"}))
}
async fn list(data: web::Data<Store>) -> Result<HttpResponse> {
    let tasks = lock(&data)?;
    Ok(HttpResponse::Ok().json(serde_json::json!({"tasks": tasks.values().collect::<Vec<_>>(), "total": tasks.len()})))
}
async fn create(body: web::Json<TaskCreate>, data: web::Data<Store>) -> Result<HttpResponse> {
    validate(&body.title, &body.description)?;
    let now = Utc::now();
    let task = Task { id: Uuid::new_v4().to_string(), title: body.title.clone(),
        description: body.description.clone(), completed: false, created_at: now, updated_at: now };
    lock(&data)?.insert(task.id.clone(), task.clone());
    Ok(HttpResponse::Created().json(task))
}
async fn get(id: web::Path<String>, data: web::Data<Store>) -> Result<HttpResponse> {
    let tasks = lock(&data)?;
    let task = tasks.get(id.as_str()).ok_or_else(|| error::ErrorNotFound("Task not found"))?;
    Ok(HttpResponse::Ok().json(task))
}
async fn replace(id: web::Path<String>, body: web::Json<TaskReplace>, data: web::Data<Store>) -> Result<HttpResponse> {
    validate(&body.title, &body.description)?;
    let mut tasks = lock(&data)?;
    let task = tasks.get_mut(id.as_str()).ok_or_else(|| error::ErrorNotFound("Task not found"))?;
    task.title = body.title.clone();
    task.description = body.description.clone();
    task.completed = body.completed;
    task.updated_at = Utc::now();
    Ok(HttpResponse::Ok().json(task.clone()))
}
async fn delete(id: web::Path<String>, data: web::Data<Store>) -> Result<HttpResponse> {
    lock(&data)?.remove(id.as_str()).ok_or_else(|| error::ErrorNotFound("Task not found"))?;
    Ok(HttpResponse::NoContent().finish())
}
async fn metrics(data: web::Data<Store>) -> Result<HttpResponse> {
    let tasks = lock(&data)?;
    let complete = tasks.values().filter(|t| t.completed).count();
    let body = format!("# TYPE task_count gauge\ntask_count {}\n# TYPE task_completed_count gauge\ntask_completed_count {}\n# TYPE task_pending_count gauge\ntask_pending_count {}\n", tasks.len(), complete, tasks.len()-complete);
    Ok(HttpResponse::Ok().content_type("text/plain; version=0.0.4; charset=utf-8").body(body))
}
async fn root() -> HttpResponse {
    HttpResponse::Ok().json(serde_json::json!({"message":"Task API", "version":"1.0.0", "docker_track":"rust", "endpoints":{"health":"/health", "tasks":"/api/tasks", "metrics":"/metrics"}}))
}
#[actix_web::main]
async fn main() -> std::io::Result<()> {
    env_logger::init_from_env(env_logger::Env::new().default_filter_or("info"));
    let data = web::Data::new(Store::new(HashMap::new()));
    HttpServer::new(move || App::new().app_data(data.clone()).wrap(Logger::default())
        .route("/", web::get().to(root))
        .route("/health", web::get().to(health))
        .route("/metrics", web::get().to(metrics))
        .route("/api/tasks", web::get().to(list))
        .route("/api/tasks", web::post().to(create))
        .route("/api/tasks/{id}", web::get().to(get))
        .route("/api/tasks/{id}", web::put().to(replace))
        .route("/api/tasks/{id}", web::delete().to(delete)))
        .shutdown_timeout(10).bind(("0.0.0.0", 8080))?.run().await
}
